framework/libra-framework/sources/ol_sources/vote_lib/multi_action.move

```diff
@@ -173,19 +199,293 @@ module ol_framework::multi_action {
+        guid_capability: account::create_guid_capability(sig),
+      });
+    };
+
+    if (!exists<Action<PropGovSigners>>(multisig_address)) {
+      move_to(sig, Action<PropGovSigners> {
+        can_withdraw: false,
+        vote: ballot::new_tracker<Proposal<PropGovSigners>>(),
+      });
+    };
+  }
+
+
+  // Private function to assist governance vote
+  fun add_offer_addresses(addr: address, proposed: vector<address>) acquires Offer {
+    let offer = borrow_global_mut<Offer>(addr);
+    let duration = epoch_helper::get_current_epoch() + DEFAULT_EPOCHS_OFFER_EXPIRE;
+    let i = 0;
+    while (i < vector::length(&proposed)) {
+      let addr = vector::borrow(&proposed, i);
+      vector::push_back(&mut offer.proposed, *addr);
+      vector::push_back(&mut offer.expiration_epoch, duration);
+      i = i + 1;
+    };
+  }
+
+  // DANGER - may forge the signer of the multisig account is necessary here
+  // TODO: remove this function after offer migration is completed
+  // Migrate a legacy account to have structure Offer in order to propose authorities changes
+  public entry fun migrate_offer(sig: &signer, multisig_address: address) {
+    // Ensure the account does not have Offer structure
+    assert!(!exists_offer(multisig_address), error::already_exists(EOFFER_ALREADY_EXISTS));
+
+    // if account is multisig, forge signer and add Offer to the multisig account
+    if (multisig_account::is_multisig(multisig_address)) {
+      // a) multisig account: ensure the signer is in the authorities list
+      assert!(is_authority(multisig_address, signer::address_of(sig)), error::permission_denied(ENOT_AUTHORIZED));
+
+      // We create the signer for the multisig account here since this is required 
+      // to add the Offer resource. 
+      // This should be safe because we check that the signer is in the authorities list.
+      // Also, after all accounts are migrated this function will be deprecated.
+      let multisig_signer = &create_signer(multisig_address); // <<< DANGER
+
+      // create Offer structure
+      let offer = construct_empty_offer();
+      move_to(multisig_signer, offer);
+    } else {
+      // b) initiated account: ensure the account is initialized with governance and add Offer to the account
+      assert!(multisig_address == signer::address_of(sig), error::permission_denied(ENOT_AUTHORIZED));
+      assert!(is_gov_init(multisig_address), error::invalid_state(EGOV_NOT_INITIALIZED));
+      let offer = construct_empty_offer();
+      move_to(sig, offer);
+    };
+  }
+
+  // Propose an offer to new authorities on the signer account 
+  // or update the expiration epoch of the existing proposed authorities.
+  // - sig: The signer proposing the offer.
+  // - proposed: The list of authorities addresses proposed.
+  // - duration_epochs: The duration in epochs before the offer expires.
+  public entry fun propose_offer(sig: &signer, proposed: vector<address>, duration_epochs: Option<u64>) acquires Offer {
```

Comment:
This function should probably be broken up into multiple private functions, and the `entry fun` be minimalistic.

Reply:
got it.